File: common/src/contract.rs
//! Contract Interface.
use std::convert::TryFrom;

use bytes::B256;
use error::Error;

use ekiden_common_api as api;

/// The unserialized representation of a contract.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Contract {
    /// Globally unique long term identifier of the Contract.
    pub id: B256,

    /// Storage service ID associated with the Contract.
    pub store_id: B256,

    /// The contract code body.
    pub code: Vec<u8>,

    // XXX: "tokens" for advertisement (PR #2), in to be specified units.
    /// The minimum stake required by the contract.
    pub minimum_bond: u64,

    pub mode_nondeterministic: bool,

    pub features_sgx: bool,

    /// Number of tokens/second of contract instance advertisement.
    pub advertisement_rate: u64,

    /// The size of the computation group this contract will be sharded out to.
    pub replica_group_size: u64,

    /// The size of the storage grou pthis contract will use.
    pub storage_group_size: u64,
    // XXX: "budget" (PR #2), in to be specified units.

    // XXX: Other critera
}

impl TryFrom<api::Contract> for Contract {
    /// try_from Converts a protobuf `common::api::Contract` into a contract.
    type Error = super::error::Error;
    fn try_from(a: api::Contract) -> Result<Self, Error> {
        let id = a.get_id();
        let id = B256::from_slice(&id);

        let sid = a.get_store_id();
        let sid = B256::from_slice(&sid);

        Ok(Contract {
            id: id,
            store_id: sid,
            code: a.get_code().to_vec(),
            minimum_bond: a.minimum_bond,
            mode_nondeterministic: a.get_mode() == api::Contract_Mode::Nondeterministic,
            features_sgx: a.get_features()
                .iter()
                .any(|f| *f == api::Contract_Features::SGX),
            advertisement_rate: a.advertisement_rate,
            replica_group_size: a.replica_group_size,
            storage_group_size: a.storage_group_size,
        })
    }
}

impl Into<api::Contract> for Contract {
    /// into Converts a contract into a protobuf `common::api::Contract` representation.
    fn into(self) -> api::Contract {
        let mut c = api::Contract::new();
        c.set_id(self.id.to_vec());
        c.set_store_id(self.store_id.to_vec());
        c.set_code(self.code);
        c.set_minimum_bond(self.minimum_bond);
        if self.mode_nondeterministic {
            c.set_mode(api::Contract_Mode::Nondeterministic);
        }
        if self.features_sgx {
            c.set_features(vec![api::Contract_Features::SGX]);
        }
        c.set_advertisement_rate(self.advertisement_rate);
        c.set_replica_group_size(self.replica_group_size);
        c.set_storage_group_size(self.storage_group_size);
        c
    }
}

File: go/common/contract/contract.go
// Package contract implements common contract routines.
package contract

import (
	"encoding/hex"
	"errors"

	"github.com/oasislabs/ekiden/go/common"
	"github.com/oasislabs/ekiden/go/common/cbor"
	"github.com/oasislabs/ekiden/go/common/crypto/signature"
	pbCommon "github.com/oasislabs/ekiden/go/grpc/common"
)

var (
	// ErrMalformedStoreID is the error returned when a storage service
	// ID is malformed.
	ErrMalformedStoreID = errors.New("contract: Malformed store ID")

	// ErrNilProtobuf is the error returned when a protobuf is nil.
	ErrNilProtobuf = errors.New("node: Protobuf is nil")

	_ cbor.Marshaler   = (*Contract)(nil)
	_ cbor.Unmarshaler = (*Contract)(nil)
)

// StoreID is a storage service ID.
// TODO: Move this to the storage package when it exists.
type StoreID [32]byte

// MarshalBinary encodes a StoreID into binary form.
func (id *StoreID) MarshalBinary() (data []byte, err error) {
	data = append([]byte{}, id[:]...)
	return
}

// UnmarshalBinary decodes a binary marshaled StoreID.
func (id *StoreID) UnmarshalBinary(data []byte) error {
	const idSize = 32

	if len(data) != idSize {
		return ErrMalformedStoreID
	}

	copy(id[:], data)
	return nil
}

// String returns a string representation of a StoreID.
func (id *StoreID) String() string {
	return hex.EncodeToString(id[:])
}

// Contract represents a contract (aka runtime).
type Contract struct {
	// ID is a globally unique long term identifier of the contract.
	ID signature.PublicKey `codec:"id"`

	// StoreID is the storage service ID associated with the contract.
	StoreID StoreID `codec:"store_id"`

	// Code is the contract code body.
	Code []byte `codec:"code"`

	// MinimumBond is the mimimum stake required by the contract.
	MinimumBond uint64 `codec:"minimum_bond"`

	// ModeNonDeterministic indicates if the contract should be executed
	// in a non-deterministic manner.
	ModeNonDeterministic bool `codec:"mode_nondeterministic"`

	// FeaturesSGX indicates if the contract requires SGX.
	FeaturesSGX bool `codec:"features_sgx"`

	// AdvertisementRate is the number of tokens/second of contract
	// instance advertisement.
	AdvertisementRate uint64 `codec:"advertisement_rate"`

	// ReplicaGroupSize is the size of the computation group.
	ReplicaGroupSize uint64 `codec:"replica_group_size"`

	// ReplicaGroupBackupSize is the size of the discrepancy resolution
	// replica group.
	ReplicaGroupBackupSize uint64 `codec:"replica_group_backup_size"`

	// ReplicaAllowedStragglers is the number of allowed stragglers.
	ReplicaAllowedStragglers uint64 `codec:"replica_allowed_stragglers"`

	// StorageGroupSize is the size of the storage group.
	StorageGroupSize uint64 `codec:"storage_group_size"`
}

// String returns a string representation of itself.
func (c *Contract) String() string {
	return "<Contract id=" + c.ID.String() + ">"
}

// Clone returns a copy of itself.
func (c *Contract) Clone() common.Cloneable {
	contractCopy := *c
	return &contractCopy
}

// FromProto deserializes a protobuf into a Contract.
func (c *Contract) FromProto(pb *pbCommon.Contract) error {
	if pb == nil {
		return ErrNilProtobuf
	}

	if c.ID == nil {
		c.ID = signature.PublicKey{}
	}

	if err := c.ID.UnmarshalBinary(pb.GetId()); err != nil {
		return err
	}

	if err := c.StoreID.UnmarshalBinary(pb.GetStoreId()); err != nil {
		return err
	}

	c.Code = append([]byte{}, pb.GetCode()...)
	c.MinimumBond = pb.GetMinimumBond()
	c.AdvertisementRate = pb.GetAdvertisementRate()
	c.ReplicaGroupSize = pb.GetReplicaGroupSize()
	c.ReplicaGroupBackupSize = pb.GetReplicaGroupBackupSize()
	c.ReplicaAllowedStragglers = pb.GetReplicaAllowedStragglers()
	c.StorageGroupSize = pb.GetStorageGroupSize()

	switch pb.GetMode() {
	case pbCommon.Contract_Nondeterministic:
		c.ModeNonDeterministic = true
	default:
		c.ModeNonDeterministic = false
	}

	c.FeaturesSGX = pbWantsSGX(pb)

	return nil
}

// ToProto serializes a Contract into a protobuf.
func (c *Contract) ToProto() *pbCommon.Contract {
	pb := new(pbCommon.Contract)
	var err error

	if pb.Id, err = c.ID.MarshalBinary(); err != nil {
		return nil
	}
	if pb.StoreId, err = c.StoreID.MarshalBinary(); err != nil {
		return nil
	}
	pb.Code = append([]byte{}, c.Code...)
	pb.MinimumBond = c.MinimumBond
	pb.AdvertisementRate = c.AdvertisementRate
	pb.ReplicaGroupSize = c.ReplicaGroupSize
	pb.ReplicaGroupBackupSize = c.ReplicaGroupBackupSize
	pb.ReplicaAllowedStragglers = c.ReplicaAllowedStragglers
	pb.StorageGroupSize = c.StorageGroupSize

	switch c.ModeNonDeterministic {
	case true:
		pb.Mode = pbCommon.Contract_Nondeterministic
	case false:
		pb.Mode = pbCommon.Contract_Deterministic
	}

	if c.FeaturesSGX {
		pb.Features = append([]pbCommon.Contract_Features{}, pbCommon.Contract_SGX)
	}

	return pb
}

// ToSignable serializes the Contract into a signature compatible byte vector.
func (c *Contract) ToSignable() []byte {
	return c.MarshalCBOR()
}

// MarshalCBOR serializes the type into a CBOR byte vector.
func (c *Contract) MarshalCBOR() []byte {
	return cbor.Marshal(c)
}

// UnmarshalCBOR deserializes a CBOR byte vector into given type.
func (c *Contract) UnmarshalCBOR(data []byte) error {
	return cbor.Unmarshal(data, c)
}

func pbWantsSGX(pb *pbCommon.Contract) bool {
	for _, f := range pb.GetFeatures() {
		if f == pbCommon.Contract_SGX {
			return true
		}
	}

	return false
}

// SignedContract is a signed blob containing a CBOR-serialized Contract.
type SignedContract struct {
	signature.Signed
}

// Open first verifies the blob signature and then unmarshals the blob.
func (s *SignedContract) Open(context []byte, contract *Contract) error { // nolint: interfacer
	return s.Signed.Open(context, contract)
}
